browser/src/layer/tile/Premult.ts

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright the Collabora Online contributors.
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+declare var L: any;
+
+namespace cool {
+
+/**
+ * Used to re-use an aligned buffer that we can un-pre-multiply
+ * un-aligned bytes from a stream into, and ensure we can also
+ * access these as 32bit pixels.
+ */
+export class UnPremult {
+	public _pixels: Uint32Array;
+	public _bytes: Uint8ClampedArray;
+
+	constructor(len: number) {
+		this.sizeArrays(len);
+	}
+
+	private sizeArrays(len: number) : void {
+		this._pixels = new Uint32Array(len);
+		this._bytes = new Uint8ClampedArray(
+			this._pixels.buffer, this._pixels.byteOffset,
+			this._pixels.byteLength);
+	}
+
+	public unpremultiply(data: Uint8Array, byteLength: number, byteOffset: number = 0) : void {
+		if (byteLength > this._bytes.length)
+		{
+			window.console.warn('Unusual - extend array to unpremultiply ' + byteLength + ' vs. ' + this._bytes.length);
+			this.sizeArrays((byteLength + 3) /4);
+		}
+		for (var o8 = 0; o8 < byteLength; o8 += 4) {
+			// premultiplied rgba -> unpremultiplied rgba
```

Comment:
in core we have a unpremultiply_table table that we initialize once, don't know if that would add any value here too

Reply:
Ooh - good idea :-) at least we have an object to put it on now.